nucypher/blockchain/eth/agents.py

```diff
@@ -321,15 +330,36 @@ def get_all_active_stakers(self, periods: int, pagination_size: Optional[int] =
             start_index: int = 0
             n_tokens: int = 0
             stakers: Dict[int, int] = dict()
-            active_stakers: Tuple[NuNits, List[List[int]]]
+            attempts = 0
             while start_index < num_stakers:
-                active_stakers = self.contract.functions.getActiveStakers(periods, start_index, pagination_size).call()
-                temp_locked_tokens, temp_stakers = active_stakers
-                # temp_stakers is a list of length-2 lists (address -> locked tokens)
-                temp_stakers_map = {address: locked_tokens for address, locked_tokens in temp_stakers}
-                n_tokens = n_tokens + temp_locked_tokens
-                stakers.update(temp_stakers_map)
-                start_index += pagination_size
+                try:
+                    attempts += 1
+                    active_stakers = self.contract.functions.getActiveStakers(periods,
+                                                                              start_index,
+                                                                              pagination_size).call()
+                except Exception as e:
+                    if 'timeout' not in str(e):
+                        # exception unrelated to pagination size and timeout
+                        raise e
+                    elif pagination_size == 1 or attempts >= 3:
+                        # we tried
+                        raise e
+                    else:
+                        # reduce pagination size and retry
+                        old_pagination_size = pagination_size
+                        pagination_size = old_pagination_size // 2
+                        if pagination_size == 0:
```

Comment:
Wouldn't that possibility be already eliminated by `if pagination_size == 1` above?

Reply:
✔️  Yep. Was overly cautious here and this check can be removed.